### PR TITLE
Cleanup flux and significance maps  combination functions 

### PR DESCRIPTION
**Pull request by QRemy**
Cleanup flux and significance maps functions combination such as they have the same api.

### CODE REVIEW

File: gammapy/estimators/utils.py [hunk 1]
```diff
@@ -779,6 +782,61 @@ def get_rebinned_axis(fluxpoint, axis_name="energy", method=None, **kwargs):
...
+        List of maps with the same geometry.
+
+    Returns
+    -------
+    results : dict
+        Dictionary with keys :
+        - "significance" : joint significance map.
+        - "df" : degree of freedom map (one norm per valid bin).
+        - "npred_excess" : summed excess map.
+        - "estimator_results" : dictionary containing the estimator results for each dataset.
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:809]: What is it exactly? is it associated to a class? if yes, can you precise it?

File: gammapy/estimators/utils.py [hunk 2]
```diff
@@ -779,6 +782,61 @@ def get_rebinned_axis(fluxpoint, axis_name="energy", method=None, **kwargs):
...
+            result["ts"] * np.sign(result["npred_excess"])
+        ).reduce_over_axes()
+        npred_excess_sum += result["npred_excess"].reduce_over_axes()
+
+    significance = Map.from_geom(geom)
+    significance.data = ts_to_sigma(ts_sum.data, df.data) * np.sign(ts_sum_sign)
+    return dict(
+        significance=significance,
+        df=df,
+        npred_excess=npred_excess_sum,
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:836]: The `estimator_results` is not present (coherence with the docstring)

File: gammapy/estimators/utils.py [hunk 3]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+    dnde_scan_axis=None,
+):
+    """Create a FluxMaps by combining a list of flux maps with the same geometry.
+
+     This assumes the flux maps are independent measurements of the same true value.
+     The GTI is stacked in the process.
+
+    Parameters
+    ----------
+    estimator : `~gammapy.estimators.ExcessMapEstimator` or `~gammapy.estimators.TSMapEstimator`
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1041]:
```suggestion
    estimator : `gammapy.estimators.ExcessMapEstimator` or `gammapy.estimators.TSMapEstimator`
```

File: gammapy/estimators/utils.py [hunk 4]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+        Excess Map Estimator or TS Map Estimator
+    dataset : `~gammapy.datasets.Datasets`
+        Datasets containing only `~gammapy.datasets.MapDataset`.
+    method : {"gaussian_errors"}
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1045]:
```suggestion
    method : dict 
```

File: gammapy/estimators/utils.py [hunk 5]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+            The product of gaussians is also a gaussian so can derive dnde, dnde_err, and ts.
+        * distrib :
+            Likelihood profile approximation assuming that probabilities distributions for
+            flux points correspond to asymmetric gaussians and for upper limits to complementary error functions.
+            Use available quantities among dnde, dnde_err, dnde_errp, dnde_errn, dnde_ul, and ts.
+        * profile :
+            Sum the likelihood profile maps.
+            The flux maps must contains the `stat_scan` maps.
+
+        Default is "gaussian_errors" which is the faster but least accurate solution,
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1057]:
```suggestion
        Default is `{"gaussian_errors"}` which is the faster but least accurate solution,
```

File: gammapy/estimators/utils.py [hunk 6]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+        "distrib"  will be more accurate if dnde_errp and dnde_errn are available,
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1058]:
```suggestion
        `{"distrib"}`  will be more accurate if dnde_errp and dnde_errn are available,
```

File: gammapy/estimators/utils.py [hunk 7]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+        "profile"  will be even more accurate if "stat_scan" is available.
+
+    reference_model : `~gammapy.modeling.models.SkyModel`, optional
+        Reference model to use for conversions.
+        Default is None and is will use the reference_model of the first FluxMaps in the list.
+
+    dnde_scan_axis : `~gammapy.maps.MapAxis`
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1065]:
```suggestion
    dnde_scan_axis : `gammapy.maps.MapAxis`
```

File: gammapy/estimators/utils.py [hunk 8]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+            Use available quantities among dnde, dnde_err, dnde_errp, dnde_errn, dnde_ul, and ts.
+        * profile :
+            Sum the likelihood profile maps.
+            The flux maps must contains the `stat_scan` maps.
+
+        Default is "gaussian_errors" which is the faster but least accurate solution,
+        "distrib"  will be more accurate if dnde_errp and dnde_errn are available,
+        "profile"  will be even more accurate if "stat_scan" is available.
+
+    reference_model : `~gammapy.modeling.models.SkyModel`, optional
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1061]:
```suggestion
    reference_model : `gammapy.modeling.models.SkyModel`, optional
```

File: gammapy/estimators/utils.py [hunk 9]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+            Likelihood profile approximation assuming that probabilities distributions for
+            flux points correspond to asymmetric gaussians and for upper limits to complementary error functions.
+            Use available quantities among dnde, dnde_err, dnde_errp, dnde_errn, dnde_ul, and ts.
+        * profile :
+            Sum the likelihood profile maps.
+            The flux maps must contains the `stat_scan` maps.
+
+        Default is "gaussian_errors" which is the faster but least accurate solution,
+        "distrib"  will be more accurate if dnde_errp and dnde_errn are available,
+        "profile"  will be even more accurate if "stat_scan" is available.
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1059]:
```suggestion
        `{"profile"}`  will be even more accurate if "stat_scan" is available.
```

File: gammapy/estimators/utils.py [hunk 10]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+        Default is None and it will be derived from the first FluxMaps in the list.
+        Used only if `method` is distrib or profile.
+
+
+    Returns
+    -------
+    results : dict
+        Dictionary with entries:
+
+                * "flux_maps" : `~gammapy.estimators.FluxMaps`
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1076]:
```suggestion
                * "flux_maps" : `gammapy.estimators.FluxMaps`
```

File: gammapy/estimators/utils.py [hunk 11]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+                * "estimator_results" : dictionary containing the estimator results for each dataset.
```
Comment (bkhelifi) [inline gammapy/estimators/utils.py:1077]: Same comment as above

File: gammapy/estimators/map/tests/test_ts.py [hunk 12]
```diff
@@ -412,6 +413,14 @@ def test_ts_map_stat_scan(fake_dataset):
...
     assert_allclose(combined_map.norm.data[success], norm[success], rtol=5e-2)
 
+    combined_results = get_combined_flux_maps(
+        estimator, [dataset, dataset.copy()], method="distrib"
+    )
+    combined_map = combined_results["flux_maps"]
+    assert len(combined_results["estimator_results"].values()) == 2
+    assert_allclose(combined_map.ts.data, 2 * ts, rtol=1e-4)
+    assert_allclose(combined_map.norm.data[success], norm[success], rtol=5e-2)
+
```
Comment (bkhelifi) [inline gammapy/estimators/map/tests/test_ts.py:423]: Do you test that all keys mentioned into the docstring are present?

File: gammapy/estimators/utils.py (hunk 2 above)
Comment (QRemy) [reply gammapy/estimators/utils.py:836]: added

File: gammapy/estimators/utils.py (hunk 11 above)
Comment (QRemy) [reply gammapy/estimators/utils.py:1077]: clarified

File: gammapy/estimators/map/tests/test_ts.py (hunk 12 above)
Comment (QRemy) [reply gammapy/estimators/map/tests/test_ts.py:423]: yes it in the existing tests for `combine_flux_maps` as `get_combined_flux_maps` is build on top of it it don't retest everything

**Comment by Astro-Kirsty** [comment]
This is not yet exposed in the documentation, do we want to do that?

File: gammapy/estimators/utils.py [hunk 13]
```diff
@@ -25,12 +25,15 @@
...
-    "get_combined_significance_maps",
+    "combine_flux_maps",
+    "combine_significance_maps",
     "estimate_exposure_reco_energy",
     "find_peaks",
     "find_peaks_in_flux_map",
     "resample_energy_edges",
     "get_rebinned_axis",
+    "get_combined_flux_maps",
+    "get_combined_significance_maps",
```
Comment (QRemy) [inline gammapy/estimators/utils.py:36]: `get_combined_significance_maps` already appears in te dev doc, now the other methods also will.

File: gammapy/estimators/utils.py [hunk 14]
```diff
@@ -779,6 +782,62 @@ def get_rebinned_axis(fluxpoint, axis_name="energy", method=None, **kwargs):
...
+    results : dict
+        Dictionary with keys :
+        - "significance" : joint significance map.
+        - "df" : degree of freedom map (one norm per valid bin).
+        - "npred_excess" : summed excess map.
+        - "estimator_results" : dictionary containing the flux maps computed for each dataset.
+
+    See also
+    --------
+    get_combined_significance_maps : same method but computing the significance maps from estimators and datasets
```
Comment (Astro-Kirsty) [inline gammapy/estimators/utils.py:813]:
```suggestion
    get_combined_significance_maps : same method but computing the significance maps from estimators and datasets.
```

**Review by Astro-Kirsty — CHANGES_REQUESTED** [review]
Thanks, just a few minor comments

File: gammapy/estimators/utils.py [hunk 15]
```diff
@@ -779,6 +782,62 @@ def get_rebinned_axis(fluxpoint, axis_name="energy", method=None, **kwargs):
     return axis_new
 
 
+def combine_significance_maps(maps):
+    """Computes excess and significance for a set of datasets.
+    The significance computation assumes that the model contains
+    one degree of freedom per valid energy bin in each dataset.
+    This method implemented here is valid under the assumption
+    that the TS in each independent bin follows a Chi2 distribution,
+    then the sum of the TS also follows a Chi2 distribution (with the sum of degree of freedom).
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/estimators/utils.py]: done

**Review by registerrier — COMMENTED** [review]
Thanks @QRemy . Looks good. See small inline suggestions/comments.

File: gammapy/estimators/utils.py [hunk 16]
```diff
@@ -973,6 +1024,87 @@ def combine_flux_maps(
...
+            Use available quantities among dnde, dnde_err, dnde_errp, dnde_errn, dnde_ul, and ts.
+        * profile :
+            Sum the likelihood profile maps.
+            The flux maps must contains the `stat_scan` maps.
+
+        Default is "gaussian_errors" which is the faster but least accurate solution,
+        "distrib"  will be more accurate if dnde_errp and dnde_errn are available,
+        "profile"  will be even more accurate if "stat_scan" is available.
+
+    reference_model : `gammapy.modeling.models.SkyModel`, optional
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/estimators/utils.py:1061]: I think @bkhelifi asked to remove the ~,  could we agree on this once for all ?

File: gammapy/estimators/utils.py [hunk 17]
```diff
@@ -841,10 +835,61 @@ def get_combined_significance_maps(estimator, datasets):
...
+        raise TypeError(
+            f"estimator type should be ExcessMapEstimator or TSMapEstimator), got {type(estimator)} instead."
+        )
+
+    results = dict()
+    for dataset in datasets:
+        result = estimator.run(dataset)
+        results[dataset.name] = result
+
+    return combine_significance_maps(list(results.values()))
```
[the remark this answers is not in the document]
Comment (QRemy) [reply gammapy/estimators/utils.py]: modified

**Comment by Astro-Kirsty** [comment]
In the 'see also' sections we actually need to keep it the with the original way. It automatically renders the link! 
Otherwise the docs will not build, due to some how like a 'circular import'.

File: gammapy/estimators/utils.py [hunk 18]
```diff
@@ -807,27 +807,21 @@ def get_combined_significance_maps(estimator, datasets):
...
-    from .map.ts import TSMapEstimator
+                * "estimator_results" : dictionary containing the flux maps computed for each dataset.
 
-    if not isinstance(estimator, (ExcessMapEstimator, TSMapEstimator)):
-        raise TypeError(
-            f"estimator type should be ExcessMapEstimator or TSMapEstimator), got {type(estimator)} instead."
-        )
+    See also
+    --------
+    `get_combined_significance_maps` : same method but computing the significance maps from estimators and datasets.
```
Comment (Astro-Kirsty) [inline gammapy/estimators/utils.py:814]: Dont want to have the quotes either. Fix for all please

```suggestion
    get_combined_significance_maps : same method but computing the significance maps from estimators and datasets.
```

**Comment by QRemy** [comment]
Now that all comments are implemented I'm going to merge this.

**Pull request merged**